azure-iot-device/azure/iot/device/iothub/pipeline/pipeline_stages_iothub_mqtt.py

```diff
@@ -97,100 +95,91 @@ def _execute_op(self, op):
                 )
             )
 
-            # make a callback that can call the user's callback after the reconnect is complete
             def on_reconnect_complete(reconnect_op, error):
                 if error:
                     logger.error(
                         "{}({}) reconnection failed.  returning error {}".format(
                             self.name, op.name, error
                         )
                     )
-                    self.send_completed_op_up(op, error=error)
                 else:
                     logger.debug(
                         "{}({}) reconnection succeeded.  returning success.".format(
                             self.name, op.name
                         )
                     )
-                    self.send_completed_op_up(op)
-
-            # save the old user callback so we can call it later.
-            old_callback = op.callback
 
             # make a callback that either fails the UpdateSasTokenOperation (if the lower level failed it),
             # or issues a ReconnectOperation (if the lower level returned success for the UpdateSasTokenOperation)
             def on_token_update_complete(op, error):
-                op.callback = old_callback
                 if error:
                     logger.error(
                         "{}({}) token update failed.  returning failure {}".format(
                             self.name, op.name, error
                         )
                     )
-                    self.send_completed_op_up(op, error=error)
                 else:
                     logger.debug(
                         "{}({}) token update succeeded.  reconnecting".format(self.name, op.name)
                     )
 
-                    self.send_op_down(
-                        pipeline_ops_base.ReconnectOperation(callback=on_reconnect_complete)
+                    # Stop completion of Token Update op, and only continue upon completion of Reconnect op
+                    op.uncomplete()
+                    worker_op = op.spawn_worker_op(
+                        worker_op_type=pipeline_ops_base.ReconnectOperation,
+                        callback=on_reconnect_complete,
```

Comment:
>on_reconnect_complete [](start = 33, length = 21)

You could remove this callback (it only logs), especially if this is the only example we have with a callback inside the kwargs on a spawn_worker_op call.

Reply:
Yeah, I had considered it but felt the log might be valuable. If you think it can be removed I'll go ahead and do it though.

---
In reply to: [346003283](https://github.com/Azure/azure-iot-sdk-python/pull/392#discussion_r346003283) [](ancestors = 346003283)